File: practice/mathematics/number_theory/identify_smith_numbers/identify_smith_numbers.py
#!/usr/bin/env python

def sum_digits(i_val):
    return sum(map(int, str(i_val)))

def sieve(i_n):
    is_prime = [True for _ in range(i_n + 1)]
    i = 2
    while i * i <= i_n:
        if is_prime[i]:
            j = i * i
            while j <= i_n:
                is_prime[j] = False
                j += i
        i += 1
    return [CNT for CNT in range(2, i_n + 1) if is_prime[CNT]]

n = int(raw_input().strip())
PRIMES = sieve(min(10000000, n + 1))

d = sum_digits(n)
s = 0

for P in PRIMES:
    if n < P:
        break
    if n % P == 0:
        t = sum_digits(P)
        while n % P == 0:
            s += t
            n //= P
    if s > d:
        break
if s == d:
    print 1
else:
    print 0

File: practice/mathematics/number_theory/identify_smith_numbers/identify_smith_numbers.cpp
#include <vector>
#include <iostream>

using namespace std;

int sum_digits(unsigned i_value)
{
    int res = 0;
    while (i_value > 0)
    {
        res += i_value % 10;
        i_value /= 10;
    }
    return res;
}

vector<unsigned> sieve(unsigned i_n)
{
    vector<bool> is_prime(i_n + 1, true);
    
    unsigned i = 2;
    while (i * i <= i_n)
    {
        if (is_prime[i])
        {
            unsigned j = i * i;
            while (j <= i_n)
            {
                is_prime[j] = false;
                j += i;
            }
        }
        i++;
    }
    
    vector<unsigned> primes;
    for (unsigned val = 2; val < i_n + 1; ++val)
    {
        if (is_prime[val])
        {
            primes.push_back(val);
        }
    }
    
    return primes;
}

int main()
{
    int t, d, s = 0;
    
    unsigned n;
    cin >> n;
    d = sum_digits(n);
    
    vector<unsigned> primes = sieve(10000000);
    
    vector<unsigned>::iterator it = primes.begin();
    for (; it != primes.end(); ++it)
    {
        unsigned p = *it;
        if (n < p)
        {
            break;
        }
        if (n % p == 0)
        {
            t = sum_digits(p);
            while (n % p == 0)
            {
                s += t;
                n /= p;
            }
        }
        if (s > d)
        {
            break;
        }
    }
    
    if (s == d)
    {
        cout << 1 << endl;
    }
    else
    {
        cout << 0 << endl;
    }
    
    return 0;
}

